docs: Add examples and undocumented callback funcs (#10756)
obsproject/obs-studio: docs/sphinx/reference-libobs-callback.rst

.. function:: void calldata_set_ptr(calldata_t *data, const char *name, void *ptr)

>    Sets a pointer parameter. This does not have to be a pointer to a pointer.

Reviewer: This seems superfluous to me.

Reviewer 2: Without concrete examples explaining why this clarification is needed, I agree.

---------------------

.. function:: void calldata_set_data(calldata_t *data, const char *name, const void *in, size_t new_size)

   Sets a parameter with any type. For common parameter types, use the functions
   above.

>    :param data: Calldata structure
>    :param name: Parameter name
>    :param in:   Pointer to the value to be stored
>    :param new_size: Size of the value to be stored

Reviewer 2: ```suggestion
   :param data:     Calldata structure
   :param name:     Parameter name
   :param in:       Pointer to the value to be stored
   :param new_size: Size of the value to be stored
```

>    :param new_size: Size of the value to be stored

Reviewer: cc @RytoEX "of value to be stored" sounds off to me, but I'm not a native English speaker. Is this correct?

Reviewer 2: This is a "correct" way of saying this, as "to be stored" makes it clear that the value is not yet stored but will be stored in the future (with this function call). Alternatives are:
* Size of the stored value (unclear if the value will be stored or is already be stored)
* Size of the value to store (probably fine)
* Size of the value that will be stored (probably fine)

---------------------

.. function:: bool calldata_get_int(const calldata_t *data, const char *name, long long *val)

>    Gets an integer parameter. See :c:func:`calldata_int` for a simpler call.

Reviewer: Maybe just this?

```suggestion
   Gets an integer parameter. See also :c:func:`calldata_int`.
```

Alternatively, explain the use cases for the respective functions in the docs for both functions. See the relevant header file:
https://github.com/obsproject/obs-studio/blob/1ecea568172979547da9b701a14655ab6f6611a2/libobs/callback/calldata.h#L124-L129

   :param data: Calldata structure
   :param name: Parameter name
   :param val:  Pointer to an integer receiving the parameter

>    :return:     ``true`` if the stored value is successfully cast

Reviewer: IMO "is successfully cast" implies more type safety guarantees than there actually are. I guess saying it does a cast is technically correct, but since it doesn't check for anything other than the size of the data (unlike e.g. an `std::variant`), I find "successful" a bit misleading.

I would just do something like this:
```suggestion
   :return:     ``true`` on success, ``false`` otherwise
```

>    :return:     ``true`` if the stored value is successfully casted

Reviewer 2: ```suggestion
   :return:     ``true`` if the stored value is successfully cast
```

---------------------

.. function:: bool calldata_get_float(const calldata_t *data, const char *name, double *val)

   Gets a floating point parameter. See :c:func:`calldata_float` for a simpler
   call.

   :param data: Calldata structure
   :param name: Parameter name
   :param val:  Pointer to a floating point receiving the parameter

>    :return:     ``true`` if the stored value is successfully casted

Reviewer 2: ```suggestion
   :return:     ``true`` if the stored value is successfully cast
```

---------------------

.. function:: bool calldata_get_bool(const calldata_t *data, const char *name, bool *val)

   Gets a boolean parameter. See :c:func:`calldata_bool` for a simpler call.

   :param data: Calldata structure
   :param name: Parameter name
   :param val:  Pointer to a boolean receiving the parameter

>    :return:     ``true`` if the stored value is successfully casted

Reviewer 2: ```suggestion
   :return:     ``true`` if the stored value is successfully cast
```

---------------------

.. function:: bool calldata_get_ptr(const calldata_t *data, const char *name, void *p_ptr)

   Gets a pointer parameter. See :c:func:`calldata_ptr` for a simpler call.

   :param data:   Calldata structure
   :param name:   Parameter name
   :param p_ptr:  Pointer to a pointer receiving the parameter

>    :return:     ``true`` if the stored value is successfully casted

Reviewer 2: ```suggestion
   :return:       ``true`` if the stored value is successfully cast
```

---------------------

.. function:: bool calldata_get_string(const calldata_t *data, const char *name, const char **str)

   Gets a string parameter. See :c:func:`calldata_string` for a simpler call.

   :param data: Calldata structure
   :param name: Parameter name
   :param str:  Pointer to a string receiving the parameter

>    :return:     ``true`` if the stored value is successfully casted

Reviewer 2: ```suggestion
   :return:     ``true`` if the stored value is successfully cast
```

   For scripting, use :py:func:`signal_handler_connect`.

>    Example connecting to ``rename`` signal from :ref:`source_signal_handler_reference`,
>    declared as ``void rename(ptr source, string new_name, string prev_name)``:
> 
>    .. code:: cpp
> 
>       /* Sample source data */
>       struct my_source {
>               obs_source_t *source;
>               char *prev_name;
>               ...
>       };
> 
>       void rename_cb(void *data, calldata_t *cd) {
>               struct my_source *ms = data;
>               obs_source_t *source = calldata_ptr(cd, "source");
>               /* We could also just access the source from our source data */
>               source = ms->source;
>               const char *new_name = calldata_string(cd, "new_name");
>               const char *prev_name = calldata_string(cd, "prev_name");
> 
>               /* Do processing here */
> 
>               /* `prev_name` will be freed after the callback, so if we want a
>                * reference to it after the callback, we must duplicate it.
>                */
>               bfree(ms->prev_name);
>               ms->prev_name = bstrdup(prev_name);
>       }
> 
>       /* Assuming `ms` already contains our source data */
>       struct my_source *ms;
>       signal_handler_t *sh = obs_source_get_signal_handler(ms->source);
>       signal_handler_connect(sh, "rename", rename_cb, ms);
> 

Reviewer: Not a fan of how long this example is. I think it should be fine to cut it down a bit:
```suggestion
   Example connecting to ``rename`` signal from :ref:`source_signal_handler_reference`:

   .. code:: cpp

      obs_source_t *source = ...;

      void rename_cb(void *data, calldata_t *cd) {
              obs_source_t *source = calldata_ptr(cd, "source");
              const char *new_name = calldata_string(cd, "new_name");
              const char *prev_name = calldata_string(cd, "prev_name");

              /* Do processing here */
      }

      signal_handler_t *sh = obs_source_get_signal_handler(source);
      signal_handler_connect(sh, "rename", rename_cb, NULL);
```

If you want to emphasize the lifetime of the `calldata_t` object, just a comment like this below `/* Do processing here */`:
```
/* Keep in mind that any pointers from ``cd`` are only valid in the scope of the callback */
```

.. function:: void signal_handler_signal(signal_handler_t *handler, const char *signal, calldata_t *params)

   Emits a signal, calling all connected callbacks.

   :param handler: Signal handler object
   :param signal:  Name of signal to emit
   :param params:  Parameters to pass to the signal

   Example:

   .. code:: cpp

      obs_source_t *source = obs_get_source_by_name("Image Slideshow Source");

      /* Declaration string: void slide_changed(int index, string path) */
      calldata_t cd = {0};

>       calldata_set_int(&cd, "index", 1);
> 		calldata_set_string(&cd, "path", "path/to/image.png");
> 
> 		signal_handler_t *sh =
> 			obs_source_get_signal_handler(source);
> 		signal_handler_signal(sh, "slide_changed", &cd);
>       calldata_free(&cd);
>       obs_source_release(source);

Reviewer 2: ```suggestion
      calldata_set_int(&cd, "index", 1);
      calldata_set_string(&cd, "path", "path/to/image.png");

      signal_handler_t *sh = obs_source_get_signal_handler(source);
      signal_handler_signal(sh, "slide_changed", &cd);
      calldata_free(&cd);
      obs_source_release(source);
```

.. function:: void proc_handler_add(proc_handler_t *handler, const char *decl_string, proc_handler_proc_t proc, void *data)

   Adds a procedure to a procedure handler. Will fail if the declaration string
   has invalid syntax or a procedure with the same name was already added. Other
   than the function identifier, the ``decl_string`` is mostly for
   documentation, as libobs does not strictly enforce how the parameters are

>    handled.

Reviewer: I would remove anything after "Will fail if the declaration string has invalid syntax or a procedure with the same name was already added.", I'd say the missing verification is more of a somewhat unfortunate implementation detail.

   :param handler:     Procedure handler object
   :param decl_string: C-style function declaration string of the procedure. The
                       function identifier dictates the procedure name.
                       Parameter types can be ``int``, ``float``, ``bool``,
                       ``ptr``, ``string``, which indicates the appropriate
                       calldata function to get the parameters. May also be
                       preceded by ``in`` or ``out`` indicating how the
                       parameters are used.
   :param proc:        Procedure callback
   :param data:        Private data to pass to the callback

   Example code:

   .. code:: cpp

>       /* Sample source data */
>       struct my_source {
>               obs_source_t *source;
>               bool active;
>               ...
>       };
> 
>       static void proc_activate(void *data, calldata_t *cd)
>       {
>               struct my_source *ms = data;
>               ms->active = calldata_bool(cd, "active");
>       }
> 
>       /* Assuming `ms` already contains our source data */
>       struct my_source *ms;
>       proc_handler_t *ph = obs_source_get_proc_handler(ms->source);
>       proc_handler_add(ph, "void activate(bool active)",
>               proc_activate, ms);

Reviewer: Shorter version:
```suggestion
      obs_source_t *source = ...;

      static void proc_activate(void *data, calldata_t *cd)
      {
              bool active = calldata_bool(cd, "active");
              /* Do processing here */
      }

      proc_handler_t *ph = obs_source_get_proc_handler(source);
      proc_handler_add(ph, "void activate(bool active)", proc_activate, NULL);
```
